File: crossecore-java/src/java_/AnnotationTypeMemberDeclarationBase.java
package java_;
import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.common.notify.NotificationChain;
import org.eclipse.emf.common.notify.impl.NotificationImpl;
import org.eclipse.emf.ecore.InternalEObject;
import org.eclipse.emf.ecore.*;
import org.eclipse.emf.ecore.impl.ENotificationImpl;
public class AnnotationTypeMemberDeclarationBase 
extends BodyDeclarationImpl implements AnnotationTypeMemberDeclaration
{
	private Expression _default;
	public Expression getDefault()
	{
		
			return _default;
		
		
	}
	public void setDefault(Expression value){
		if (value != _default) {
			NotificationChain msgs = null;
			if (_default != null){
				msgs = ((InternalEObject)_default).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT, null, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT, null, msgs);
			}
			msgs = basicSetDefault(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT , value, value));
		}
	}
	private TypeAccess _type;
	public TypeAccess getType()
	{
		
			return _type;
		
		
	}
	public void setType(TypeAccess value){
		if (value != _type) {
			NotificationChain msgs = null;
			if (_type != null){
				msgs = ((InternalEObject)_type).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE, null, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE, null, msgs);
			}
			msgs = basicSetType(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE , value, value));
		}
	}
	private com.crossecore.ocl.Set<AnnotationMemberValuePair> _usages;
	
	public com.crossecore.ocl.Set<AnnotationMemberValuePair> getUsages()
	{
		if(_usages==null){
			_usages = new com.crossecore.ocl.Set<AnnotationMemberValuePair>(AnnotationMemberValuePair.class, this, Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES, Java_PackageImpl.ANNOTATION_MEMBER_VALUE_PAIR__MEMBER);
		}
		return _usages;
	
	}

	
	@Override
	protected EClass eStaticClass() {
		return Java_PackageImpl.Literals.ANNOTATIONTYPEMEMBERDECLARATION;
	}
	
	@Override
	public NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES:
				return getUsages().basicAdd((AnnotationMemberValuePair)otherEnd, msgs);
		}
		return super.eInverseAdd(otherEnd, featureID, msgs);
	}
	
	@Override
	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES:
				return getUsages().basicRemove((AnnotationMemberValuePair)otherEnd, msgs);
		}
		return super.eInverseRemove(otherEnd, featureID, msgs);
	}
	
	public NotificationChain basicSetDefault(Expression newobj, NotificationChain msgs) {
		Expression oldobj = _default;
		_default = newobj;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT, oldobj, newobj);
			if (msgs == null){
				msgs = notification;
			}
			else{
				msgs.add(notification);
			}
		}
		return msgs;
	}
	public NotificationChain basicSetType(TypeAccess newobj, NotificationChain msgs) {
		TypeAccess oldobj = _type;
		_type = newobj;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE, oldobj, newobj);
			if (msgs == null){
				msgs = notification;
			}
			else{
				msgs.add(notification);
			}
		}
		return msgs;
	}
	
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT:
				return getDefault();
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE:
				return getType();
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES:
				return getUsages();
		}
		return super.eGet(featureID, resolve, coreType);
	}
	
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT:
				setDefault((Expression) newValue);
				return;
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE:
				setType((TypeAccess) newValue);
				return;
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES:
				getUsages().clear();
				getUsages().addAll((java.util.Collection<? extends AnnotationMemberValuePair>) newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}
	
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__DEFAULT:
				return _default != null; //single != null;
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__TYPE:
				return _type != null; //single != null;
			case Java_PackageImpl.ANNOTATION_TYPE_MEMBER_DECLARATION__USAGES:
				return getUsages().isEmpty(); //many, volatile
		}
		return super.eIsSet(featureID);
	}
	
	
	
}

File: crossecore-csharp/Java/Java/Java_/AnnotationTypeMemberDeclarationBase.cs
/* CrossEcore is a cross-platform modeling framework that generates C#, TypeScript, 
 * JavaScript, Swift code from Ecore models with embedded OCL (http://www.crossecore.org/).
 * The original Eclipse Modeling Framework is available at https://www.eclipse.org/modeling/emf/.
 * 
 * contributor: Simon Schwichtenberg
 */
 
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using oclstdlib;
using Ecore;
namespace Java_{
	public class AnnotationTypeMemberDeclarationBase 
	:BodyDeclarationImpl, AnnotationTypeMemberDeclaration
	{
		private Expression _default;
		public virtual Expression default_
		{
			get {
			
				return _default;
			}
			set {
				if (value != _default) {
					NotificationChain msgs = null;
					if (_default != null){
						msgs = ((InternalEObject)_default).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT, null, msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT, null, msgs);
					}
					msgs = basicSetDefault(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT , value, value));
				}
				}
		}
		private TypeAccess _type;
		public virtual TypeAccess type
		{
			get {
			
				return _type;
			}
			set {
				if (value != _type) {
					NotificationChain msgs = null;
					if (_type != null){
						msgs = ((InternalEObject)_type).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE, null, msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE, null, msgs);
					}
					msgs = basicSetType(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE , value, value));
				}
				}
		}
		private Set<AnnotationMemberValuePair> _usages;
		
		public virtual Set<AnnotationMemberValuePair> usages
		{
			get {
				if(_usages==null){
					_usages = new Set<AnnotationMemberValuePair>(this, Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES, Java_PackageImpl.ANNOTATIONMEMBERVALUEPAIR_MEMBER);
				}
				return _usages;
			}
		
		}
		
		protected override EClass eStaticClass() {
			return Java_PackageImpl.Literals.ANNOTATIONTYPEMEMBERDECLARATION;
		}
		
		public override NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES:
					return usages.basicAdd((AnnotationMemberValuePair)otherEnd, msgs);
			}
			return base.eInverseAdd(otherEnd, featureID, msgs);
		}
		
		public override NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES:
					return usages.basicRemove((AnnotationMemberValuePair)otherEnd, msgs);
			}
			return base.eInverseRemove(otherEnd, featureID, msgs);
		}
		
		public NotificationChain basicSetDefault(Expression newobj, NotificationChain msgs) {
			var oldobj = _default;
			_default = newobj;
			if (eNotificationRequired()) {
				var notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT, oldobj, newobj);
				if (msgs == null){
					msgs = notification;
				}
				else{
					msgs.add(notification);
				}
			}
			return msgs;
		}
		public NotificationChain basicSetType(TypeAccess newobj, NotificationChain msgs) {
			var oldobj = _type;
			_type = newobj;
			if (eNotificationRequired()) {
				var notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE, oldobj, newobj);
				if (msgs == null){
					msgs = notification;
				}
				else{
					msgs.add(notification);
				}
			}
			return msgs;
		}
		
		public override object eGet(int featureID, bool resolve, bool coreType) {
			switch (featureID) {
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT:
					return default_;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE:
					return type;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES:
					return usages;
			}
			return base.eGet(featureID, resolve, coreType);
		}
		
		
		public override void eSet(int featureID, object newValue) {
			switch (featureID) {
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT:
					default_ = (Expression) newValue;
					return;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE:
					type = (TypeAccess) newValue;
					return;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES:
					usages.Clear();
					usages.AddRange(((List<EObject>)newValue)?.Cast<AnnotationMemberValuePair>());
					return;
			}
			base.eSet(featureID, newValue);
		}
		
		/*
		public override bool eIsSet(int featureID) {
			switch (featureID) {
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_DEFAULT:
					return _default != null; //single != null;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_TYPE:
					return _type != null; //single != null;
				case Java_PackageImpl.ANNOTATIONTYPEMEMBERDECLARATION_USAGES:
					return getUsages().isEmpty(); //many, volatile
			}
			return base.eIsSet(featureID);
		}
		*/
		
		
		
	}
}
